File: cpp/day5.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <regex>
#include <functional>
#include <optional>
#include <algorithm>
#include <numeric>
#include <limits>
#include <ranges>

class Day5 {
private:
    static inline const std::regex number{R"((\d+))"};
    
public:
    static std::vector<long> makeSeq(std::vector<std::string>& input, size_t& pos) {
        std::vector<long> result;
        std::sregex_iterator begin(input[pos].begin(), input[pos].end(), number);
        std::sregex_iterator end;
        
        for (auto it = begin; it != end; ++it) {
            result.push_back(std::stol(it->str()));
        }
        
        pos += 2; // skip current line and blank line
        return result;
    }
    
    static std::optional<std::function<long(long)>> makeMap(std::vector<std::string>& input, size_t& pos) {
        if (pos >= input.size()) {
            return std::nullopt;
        }
        
        pos++; // skip section header
        
        std::vector<std::tuple<long, long, long>> ranges;
        
        while (pos < input.size() && !input[pos].empty()) {
            std::sregex_iterator begin(input[pos].begin(), input[pos].end(), number);
            std::sregex_iterator end;
            
            std::vector<long> numbers;
            for (auto it = begin; it != end; ++it) {
                numbers.push_back(std::stol(it->str()));
            }
            
            if (numbers.size() >= 3) {
                ranges.emplace_back(numbers[0], numbers[1], numbers[2]);
            }
            pos++;
        }
        pos++; // skip blank line after ranges
        
        return std::function<long(long)>([ranges](long i) -> long {
            for (const auto& [b, s, l] : ranges) {
                if (i >= s && i < s + l) {
                    return b + i - s;
                }
            }
            return i;
        });
    }
    
    static std::pair<long, long> process(std::vector<std::string>& input) {
        size_t pos = 0;
        auto seeds = makeSeq(input, pos);
        
        std::vector<std::function<long(long)>> transformations;
        
        while (auto mapper = makeMap(input, pos)) {
            transformations.push_back(*mapper);
        }
        
        // Compose all transformations into a single function
        auto seedToLocation = [transformations](long seed) -> long {
            return std::accumulate(transformations.begin(), transformations.end(), seed,
                [](long value, const auto& transform) {
                    return transform(value);
                });
        };
        
        // Part 1: Transform individual seeds
        long part1 = std::numeric_limits<long>::max();
        for (long seed : seeds) {
            part1 = std::min(part1, seedToLocation(seed));
        }
        
        // Part 2: Transform seed ranges
        long part2 = std::numeric_limits<long>::max();
        for (size_t i = 0; i < seeds.size(); i += 2) {
            long start = seeds[i];
            long length = seeds[i + 1];
            
            for (long seed = start; seed < start + length; ++seed) {
                part2 = std::min(part2, seedToLocation(seed));
            }
        }
        
        return {part1, part2};
    }
    
    static void main() {
        std::vector<std::string> input;
        std::string line;
        
        while (std::getline(std::cin, line)) {
            input.push_back(line);
        }
        
        auto [part1, part2] = process(input);
        std::cout << "(" << part1 << ", " << part2 << ")" << std::endl;
    }
};

int main() {
    Day5::main();
    return 0;
}

File: csharp/Day5.cs
﻿using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;


public class Day5
{
    public static void Main(string[] args)
    {
        var input = ReadStdinLines();
        var result = Process(input);

        Console.WriteLine(result.Part1); // 174137457
        Console.WriteLine(result.Part2); // 1493866
    }

    static IEnumerator<string> ReadStdinLines()
    {
        var lines = new List<string>();
        string? line;
        while ((line = Console.ReadLine()) != null)
        {
            lines.Add(line);
        }
        return lines.GetEnumerator();
    }
    
    private static readonly Regex number = new Regex(@"(\d+)", RegexOptions.Compiled);

    static List<long> GetNumbersFromLine(string line)
    {
        return number.Matches(line)
            .Select(match => long.Parse(match.Groups[1].Value))
            .ToList();
    }

    static List<long> MakeSeq(IEnumerator<string> stream)
    {   
        stream.MoveNext();
        var result = GetNumbersFromLine(stream.Current);
	stream.MoveNext(); // skip blank line
        return result;
    }

    record Triple(long Base, long Start, long Length) {}

    static Func<long, long>? MakeMap(IEnumerator<string> stream)
    {
        if (!stream.MoveNext()) return null;

        var ranges = new List<Triple>();
        string line;

        while (stream.MoveNext()) {
            line = stream.Current;
            if (string.IsNullOrEmpty(line)) break;
            var nums = GetNumbersFromLine(line);
            Console.Error.WriteLine(string.Join(" ", nums));
            ranges.Add(new Triple(nums[0], nums[1], nums[2]));
        }

        Console.Error.WriteLine(ranges.Last());

        long TheMap(long i)
        {
            var range = ranges.FirstOrDefault(r => r.Start <= i && i < r.Start + r.Length);
            if (range == null) return i;
            return range.Base + i - range.Start;
        };

        return TheMap;
    }

    record Result(long Part1, long Part2) {}

    static Result Process(IEnumerator<string> input)
    {
        var seeds = MakeSeq(input);

        Console.Error.WriteLine("seeds read: " + string.Join(" ", seeds));

        IEnumerable<Func<long, long>> MakeMaps() {
            while (true) {
                var map = MakeMap(input);
                if (map == null) break;
                yield return map;
            }
        }

        var seedToLocation = MakeMaps().Aggregate((f, g) => s => g(f(s)));

        var part1 = seeds
            .Select(seedToLocation)
            .Min();
        Console.Error.WriteLine($"part1: {part1}");

        IEnumerable<long> CreateRange(long start, long end) {
            for (long i = start; i < end; i++) {
                yield return i;
            }
        }

        var part2 = seeds
            .Chunk(2)
            .Select(p =>
                CreateRange(p[0], p[0] + p[1])
                    .Select(seedToLocation)
                    .Min()
            )
            .Min();
        Console.Error.WriteLine($"part2: {part2}");

        return new Result(part1, part2);
    }
}
